src/hotspot/share/opto/predicates.hpp

```diff
@@ -973,6 +973,36 @@ class NodeInOriginalLoopBody : public NodeInLoopBody {
   }
 };
 
+// This class checks whether a node is in the main loop body and not the pre loop body. We cannot use the
+// NodeInOriginalLoopBody class because PhaseIdealLoop::clone_up_backedge_goo() could clone additional nodes that
+// should be pinned at the main loop body entry. The check in NodeInOriginalLoopBody will ignore these.
+class NodeInMainLoopBody : public NodeInLoopBody {
+  const uint _first_node_index_in_pre_loop_body;
+  const uint _last_node_index_in_pre_loop_body;
+  const Node_List& _old_new;
+
+  public:
+  NodeInMainLoopBody(const uint first_node_index_in_pre_loop_body, const uint last_node_index_in_pre_loop_body,
+                 const Node_List& old_new)
+      : _first_node_index_in_pre_loop_body(first_node_index_in_pre_loop_body),
+        _last_node_index_in_pre_loop_body(last_node_index_in_pre_loop_body),
+        _old_new(old_new) {}
+  NONCOPYABLE(NodeInMainLoopBody);
+
+  // Check if 'node' is not a cloned node (i.e. "< _first_node_index_in_cloned_loop_body") and if we've created a
+  // clone from 'node' (i.e. _old_new entry is non-null). Then we know that 'node' belongs to the original loop body.
+  // Additionally check if a node was cloned after the pre loop was created. This indicates that it was created by
+  // PhaseIdealLoop::clone_up_backedge_goo(). These nodes should also be pinned at the main loop entry.
+  bool check(Node* node) const override {
```

Comment:
Can it be more meaningful method's name? And I did not find where it is used.

Reply:
Since the interface name is `NodeInLoopBody`, I went with `NodeInLoopBody::check()`. But could also rename it to `check_node_in_loop_body()` for better readability.

It is used here:
https://github.com/openjdk/jdk/blob/06ff4c17bdfab9e7aa323425cb18dfdc963aa972/src/hotspot/share/opto/predicates.cpp#L157

Reply:
I was confused because it was not use in changes but I now see that it is virtual method.